libcudacxx/test/libcudacxx/cuda/memory_resource/cuda_memory_resource/allocate.pass.cpp

```diff
@@ -0,0 +1,50 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11
+// UNSUPPORTED: nvrtc
+
+#include <cuda/memory_resource>
+#include <cuda/std/cassert>
+#include <cuda/std/cstdint>
+#include <cuda/stream_ref>
+
+void ensure_device_ptr(void* ptr) {
+    assert(ptr != nullptr);
+    cudaPointerAttributes attributes;
+    cudaError_t status = cudaPointerGetAttributes (&attributes, ptr);
+    assert(status == cudaSuccess);
+    assert(attributes.type == cudaMemoryTypeDevice );
+}
+
+void test() {
+  cuda::mr::cuda_memory_resource res{};
+
+  { // allocate / deallocate
+    auto* ptr = res.allocate(42);
+    static_assert(cuda::std::is_same<decltype(ptr), void*>::value, "");
+    ensure_device_ptr(ptr);
+
+    res.deallocate(ptr, 42);
+  }
+
+  { // allocate / deallocate with alignment
+    auto* ptr = res.allocate(42, 4);
```

Comment:
Suggestion: check that the returned pointer satisfies requested alignment.
Suggestion: test unsupported alignment throws (if you change assert to throw)?

Reply:
The returned pointer wont satisfy the requested alignment unless the default alignment was passed in. 

So I would rather not test cudaMalloc

I added two tests to see what happens if we throw exceptions. But I am not aware of any configuration that actually uses exceptions

Reply:
> The returned pointer wont satisfy the requested alignment unless the default alignment was passed in.

It _better_ satisfy the requested alignment! That's the whole point of the parameter!

To be clear, it's completely fine if the pointer has a _larger_ alignment than what was requested. The check should be for `actual_alignment >= requested_alignment`.

Reply:
> It better satisfy the requested alignment! That's the whole point of the parameter!

Begging the question?

> The check should be for `actual_alignment >= requested_alignment`.

Yep, that's what I'm suggesting.